api/src/main/java/com/epam/pipeline/manager/metadata/MetadataManager.java

```diff
@@ -402,4 +428,59 @@ private void checkEntityCanBeModified(final Object entity) {
                 .ifPresent(tool -> Assert.isTrue(tool.isNotSymlink(), messageHelper.getMessage(
                         MessageConstants.ERROR_TOOL_SYMLINK_MODIFICATION_NOT_SUPPORTED)));
     }
+
+    private Map<String, String> resolveInstanceTagsFromMetadata(final String dockerImage,
+                                                                final Map<String, String> customTags) {
+        final Set<String> instanceTagsKeys = new HashSet<>(Arrays.asList(preferenceManager.findPreference(
+                SystemPreferences.CLUSTER_INSTANCE_ALLOWED_CUSTOM_TAGS)
+                .filter(StringUtils::isNotBlank)
+                .map(value -> value.split(","))
+                .orElse(Strings.EMPTY_ARRAY)));
+        if (CollectionUtils.isEmpty(instanceTagsKeys)) {
+            return customTags;
+        }
+
+        final Tool tool = toolManager.loadByNameOrId(dockerImage);
+        final MetadataEntry toolMetadata = loadMetadataItem(tool.getId(), AclClass.TOOL);
+
+        final Map<String, PipeConfValue> metadataData = MapUtils.emptyIfNull(Objects.isNull(toolMetadata)
+                ? null
+                : toolMetadata.getData());
```

Comment:
Optionals are perfect for such a situation.

```suggestion
        final Map<String, PipeConfValue> metadataData = Optional.ofNullable(toolMetadata)
                .map(MetadataEntry::getData)
                .orElseGet(Collections::emptyMap);
```